processor/lsmintervalprocessor/internal/merger/limits/tracker.go

```diff
@@ -0,0 +1,149 @@
+	// though, overflow buckets can have overflow of their own.
+	observedCount  int64
+	overflowCounts *hyperloglog.Sketch
+}
+
+func NewTracker(maxCardinality int64) *Tracker {
+	return &Tracker{maxCardinality: maxCardinality}
+}
+
+func (t *Tracker) HasOverflow() bool {
+	return t.overflowCounts != nil
+}
+
+func (t *Tracker) EstimateOverflow() uint64 {
+	if t.overflowCounts == nil {
+		return 0
+	}
+	return t.overflowCounts.Estimate()
+}
+
+// CheckOverflow checks if overflow will happen on addition of a new entry with
+// the provided hash denoting the entries ID. It assumes that any entry passed
+// to this method is a NEW entry and the check for this is left to the caller.
+func (t *Tracker) CheckOverflow(hash uint64) bool {
+	if t.maxCardinality == 0 {
+		return false
+	}
+	if t.observedCount == t.maxCardinality {
+		if t.overflowCounts == nil {
+			// Creates an overflow with 14 precision
+			t.overflowCounts = hyperloglog.New14()
+		}
+		t.overflowCounts.InsertHash(hash)
+		return true
+	}
+	t.observedCount++
+	return false
+}
+
+// MergeEstimators merges the overflow estimators for the two trackers.
+// Note that other required maintenance of the tracker for merge needs to
+// done by the caller.
+func (t *Tracker) MergeEstimators(other *Tracker) error {
+	if other.overflowCounts == nil {
+		// nothing to merge
+		return nil
+	}
+	if t.overflowCounts == nil {
+		t.overflowCounts = other.overflowCounts.Clone()
+		return nil
+	}
+	return t.overflowCounts.Merge(other.overflowCounts)
+}
+
+// MarshalWithPrefix marshals the tracker with a prefix. To be used to encode
+// more than one limit in an attribute map.
+func (t *Tracker) MarshalWithPrefix(p string, m pcommon.Map) error {
+	m.PutInt(p+counterKey, t.observedCount)
+	if t.overflowCounts != nil {
+		hll, err := t.overflowCounts.MarshalBinary()
+		if err != nil {
+			return fmt.Errorf("failed to marshal limits: %w", err)
+		}
+		v := m.PutEmptyBytes(p + hllSketchKey)
+		v.FromRaw(hll)
+	}
+	return nil
+}
+
+// Marshal encodes the tracker as attributes in the provided map.
+func (t *Tracker) Marshal(m pcommon.Map) error {
+	return t.MarshalWithPrefix("", m)
+}
+
+// UnmarshalWithPrefix unmarshals the tracker encoded with a prefix.
+func (t *Tracker) UnmarshalWithPrefix(p string, m pcommon.Map) error {
+	var errs []error
+	prefixCounterKey := p + counterKey
+	prefixHllKey := p + hllSketchKey
+	m.RemoveIf(func(k string, v pcommon.Value) bool {
+		switch k {
+		case prefixCounterKey:
+			t.observedCount = v.Int()
+			return true
+		case prefixHllKey:
+			t.overflowCounts = hyperloglog.New14()
+			if err := t.overflowCounts.UnmarshalBinary(v.Bytes().AsRaw()); err != nil {
+				errs = append(errs, fmt.Errorf(
+					"failed to unmarshal overflow estimator hll sketch: %w", err,
+				))
+			}
+			return true
+		}
+		return false
+	})
+	if len(errs) > 0 {
+		return fmt.Errorf("failed to unmarshal tracker: %w", errors.Join(errs...))
+	}
+	return nil
```

Comment:
Should we check that we found the expected keys here?

Reply:
Good point, the prefix counter key must be present if the tracker was decoded so it should be an error if tracker was expected and not found. Will add this.